selfservice/strategy/oidc/provider_lark.go

```diff
@@ -0,0 +1,176 @@
+// Copyright © 2022 Ory Corp
+// SPDX-License-Identifier: Apache-2.0
+
+package oidc
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"net/url"
+	"time"
+
+	"github.com/hashicorp/go-retryablehttp"
+	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+
+	"github.com/ory/herodot"
+	"github.com/ory/x/httpx"
+)
+
+type ProviderLark struct {
+	config *Configuration
+	reg    dependencies
+}
+
+func NewProviderLark(
+	config *Configuration,
+	reg dependencies,
+) *ProviderLark {
+	return &ProviderLark{
+		config: config,
+		reg:    reg,
+	}
+}
+
+func (g *ProviderLark) Config() *Configuration {
+	return g.config
+}
+
+func (g *ProviderLark) OAuth2(ctx context.Context) (*oauth2.Config, error) {
+	var endpoint = oauth2.Endpoint{
+		AuthURL:   "https://passport.feishu.cn/suite/passport/oauth/authorize",
+		TokenURL:  "https://passport.feishu.cn/suite/passport/oauth/token",
+		AuthStyle: oauth2.AuthStyleInParams,
+	}
+
+	return &oauth2.Config{
+		ClientID:     g.config.ClientID,
+		ClientSecret: g.config.ClientSecret,
+		Endpoint:     endpoint,
+		// DingTalk only allow to set scopes: openid or openid corpid
+		Scopes:      g.config.Scope,
+		RedirectURL: g.config.Redir(g.reg.Config().OIDCRedirectURIBase(ctx)),
+	}, nil
+
+}
+
+func (g *ProviderLark) Claims(ctx context.Context, exchange *oauth2.Token, query url.Values) (*Claims, error) {
+	type larkClaim struct {
+		Sub          string `json:"sub"`
+		Name         string `json:"name"`
+		Picture      string `json:"picture"`
+		OpenID       string `json:"open_id"`
+		UnionID      string `json:"union_id"`
+		EnName       string `json:"en_name"`
+		TenantKey    string `json:"tenant_key"`
+		AvatarURL    string `json:"avatar_url"`
+		AvatarThumb  string `json:"avatar_thumb"`
+		AvatarMiddle string `json:"avatar_middle"`
+		AvatarBig    string `json:"avatar_big"`
+		Email        string `json:"email"`
+		UserID       string `json:"user_id"`
+		Mobile       string `json:"mobile"`
+	}
+	var (
+		userEndpoint = "https://passport.feishu.cn/suite/passport/oauth/userinfo"
+		accessToken  = exchange.AccessToken
+		client       = g.reg.HTTPClient(ctx, httpx.ResilientClientDisallowInternalIPs())
+		user         larkClaim
+	)
+
+	req, err := retryablehttp.NewRequest("GET", userEndpoint, nil)
+	if err != nil {
+		return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("%s", err))
+	}
+
+	req.Header.Add("Authorization", "Bearer "+accessToken)
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("%s", err))
+	}
+	defer resp.Body.Close()
+
+	if err := json.NewDecoder(resp.Body).Decode(&user); err != nil {
+		return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("%s", err))
+	}
+
+	return &Claims{
+		Issuer:      userEndpoint,
+		Subject:     user.OpenID,
```

Comment:
Can you point to some docs that show that this is the correct ID?

Reply:
Sure, I have added a document URL at the struct definition.